drivers/ethernet/eth_smsc911x.c

```diff
@@ -0,0 +1,718 @@
+
+	smsc_read_mac_address(mac);
+
+	SMSC9220->INT_EN |= BIT(SMSC9220_INTERRUPT_RXSTATUS_FIFO_LEVEL);
+
+	net_if_set_link_addr(iface, mac, sizeof(mac), NET_LINK_ETHERNET);
+
+	context->iface = iface;
+
+	ethernet_init(iface);
+}
+
+static int smsc_write_tx_fifo(const u8_t *buf, u32_t len, bool is_last)
+{
+	u32_t *buf32;
+
+	__ASSERT_NO_MSG(((uintptr_t)buf & 3) == 0);
+
+	if (is_last) {
+		/* Last fragment may be not full */
+		len = (len + 3) & ~3;
+	}
+
+	if ((len & 3) != 0 || len == 0) {
+		LOG_ERR("Chunk size not aligned: %u", len);
+		return -1;
+	}
+
+	buf32 = (u32_t *)buf;
+	len /= 4;
+	do {
+		SMSC9220->TX_DATA_PORT = *buf32++;
+	} while (--len);
+
+	return 0;
+}
+
+static int eth_tx(struct device *dev, struct net_pkt *pkt)
+{
+	int res;
+	u16_t total_len = net_pkt_get_len(pkt);
+	u16_t sent = 0;
+	static u8_t tx_buf[1540] __aligned(4);
+	u32_t txcmd_a, txcmd_b;
+	u32_t tx_stat;
+
+	txcmd_a = (1/*is_first_segment*/ << 13) | (1/*is_last_segment*/ << 12)
+		  | total_len;
+	/* Use len as a tag */
+	txcmd_b = total_len << 16 | total_len;
+	SMSC9220->TX_DATA_PORT = txcmd_a;
+	SMSC9220->TX_DATA_PORT = txcmd_b;
+
+#define LINEARIZE
+#ifdef LINEARIZE
+	sent = net_frag_linearize(tx_buf, sizeof(tx_buf), pkt,
+				  0, sizeof(tx_buf));
+	res = smsc_write_tx_fifo(tx_buf, sent, true);
+	if (res < 0) {
+		goto error;
+	}
+#else
+	/* Currently broken, as size of net_buf's is no longer word-aligned,
+	 * e.g. first bug contains 14-bytes eth header.
+	 */
+	frag = pkt->frags;
+	while (frag) {
+		res = smsc_write_tx_fifo(frag->data, frag->len,
+					 frag->frags == NULL);
+		if (res < 0) {
+			goto error;
+		}
+		sent += frag->len;
+		frag = frag->frags;
+	}
+#endif
+
+	tx_stat = SMSC9220->TX_STAT_PORT;
+	LOG_DBG("TX_STAT: %x", tx_stat);
+
+	return 0;
+
+error:
+	LOG_ERR("Writing pkt to FIFO failed");
+	return -1;
+}
+
+static const struct ethernet_api api_funcs = {
+	.iface_api.init = eth_initialize,
+
+	.get_capabilities = eth_smsc911x_get_capabilities,
+	.send = eth_tx,
+#if defined(CONFIG_NET_STATISTICS_ETHERNET)
+	.get_stats = get_stats,
+#endif
+};
+
+static void smsc_discard_pkt(void)
+{
+	/* TODO: */
+	/* Datasheet p.43: */
+	/* When performing a fast-forward, there must be at least 4 DWORDs
+	 * of data in the RX data FIFO for the packet being discarded. For
+	 * less than 4 DWORDs do not use RX_FFWD. In this case data must be
+	 * read from the RX data FIFO and discarded using standard PIO read
+	 * operations.
+	 */
+	SMSC9220->RX_DP_CTRL = RX_DP_CTRL_RX_FFWD;
+}
+
+static inline void smsc_wait_discard_pkt(void)
+{
+	while ((SMSC9220->RX_DP_CTRL & RX_DP_CTRL_RX_FFWD) != 0) {
+	}
+}
+
+static void smsc_read_rx_fifo(u8_t *buf, u32_t len)
+{
+	u32_t *buf32;
+
+	__ASSERT_NO_MSG(((uintptr_t)buf & 3) == 0);
+	__ASSERT_NO_MSG((len & 3) == 0 && len >= 4);
+
+	buf32 = (u32_t *)buf;
+	len /= 4;
+	do {
+		*buf32++ = SMSC9220->RX_DATA_PORT;
+	} while (--len);
+}
+
+static struct net_pkt *smsc_recv_pkt(u32_t pkt_size)
+{
+		u32_t rem_size;
```

Comment:
all the code in this function seems be intended one tab too many

Reply:
> all the code in this function seems be intended one tab too many

Indeed, fixed.